src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationPerformanceMonitor.cs

```diff
@@ -2,89 +2,156 @@
 
-            if (currentPeriod > 1)
+            List<Tuple<string, int>> counts = new List<Tuple<string, int>>();
+            foreach(var componentPair in eventCounts.ToList())
             {
-                int targetPeriodSlot = Math.Min(currentPeriod, PeriodsToAverageOver - 1);
-                message.Write(FeatureCount);
-                for (int i = 0; i < FeatureCount; i++)
+                foreach(var eventPair in componentPair.Value.ToList())
                 {
-                    float spentTimeDelta = previousSpentTimes[0][i] - previousSpentTimes[targetPeriodSlot][i];
-                    float actualTimeDelta = previousActualTimes[0][i] - previousActualTimes[targetPeriodSlot][i];
-                    message.Write(spentTimeDelta / actualTimeDelta);
+                    counts.Add(new Tuple<string, int>($"{componentPair.Key}.{eventPair.Key}", eventPair.Value));
+                    eventCounts[componentPair.Key][eventPair.Key] = 0;
                 }
             }
-            else
+
+            message.Write(counts.Count);
+            foreach(var count in counts)
+            {
+                message.Write(count.Item1);
+                message.Write(count.Item2);
+            }
+        }
+
+        public static void ReadMessage(BinaryReader reader, out bool performanceMonitoringEnabled, out List<Tuple<string, double>> durations, out List<Tuple<string, int>> counts)
+        {
+            performanceMonitoringEnabled = reader.ReadBoolean();
+            durations = null;
+            counts = null;
+
+            if (!performanceMonitoringEnabled)
             {
-                message.Write(FeatureCount);
-                for (int i = 0; i < FeatureCount; i++)
+                return;
+            }
+
+            int durationsCount = reader.ReadInt32();
+            if (durationsCount > 0)
+            {
+                durations = new List<Tuple<string, double>>();
+                for (int i = 0; i < durationsCount; i++)
                 {
-                    message.Write(0.0f);
+                    string eventName = reader.ReadString();
+                    double eventDuration = reader.ReadDouble();
+                    durations.Add(new Tuple<string, double>(eventName, eventDuration));
                 }
             }
+
+            int countsCount = reader.ReadInt32();
+            if (countsCount > 0)
+            {
+                counts = new List<Tuple<string, int>>();
+                for (int i = 0; i < countsCount; i++)
+                {
+                    string eventName = reader.ReadString();
+                    int eventCount = reader.ReadInt32();
+                    counts.Add(new Tuple<string, int>(eventName, eventCount));
+                }
+            }
+        }
+
+        public void SetDiagnosticMode(bool enabled)
+        {
+            StateSynchronizationPerformanceParameters.EnablePerformanceReporting = enabled;
         }
 
         private struct TimeScope : IDisposable

```

Comment:
If we are on 7.3, we can use ref struct, I believe

Reply:
i'm going to go ahead and define this as a struct for now in case more variables exist in the future